src/Core/Signers/SignerV4.php

```diff
@@ -73,49 +94,148 @@ public function sign(Request $request, ?Credentials $credentials): void
-
-        $signingKey = 'AWS4' . $credentials->getSecretKey();
-        foreach ($credentialScope as $scopePart) {
-            $signingKey = hash_hmac('sha256', $scopePart, $signingKey, true);
-        }
-
         $stringToSign = implode(
             "\n",
             [
-                self::ALGORITHM,
+                self::ALGORITHM_REQUEST,
                 $amzDate,
                 implode('/', $credentialScope),
                 hash('sha256', $canonicalRequest),
             ]
         );
 
+        $signature = hash_hmac('sha256', $stringToSign, $signingKey);
+
         $authorizationHeader = sprintf(
             '%s Credential=%s/%s, SignedHeaders=%s, Signature=%s',
-            self::ALGORITHM,
+            self::ALGORITHM_REQUEST,
             $credentials->getAccessKeyId(),
             implode('/', $credentialScope),
             implode(';', \array_keys($canonicalHeaders)),
-            hash_hmac('sha256', $stringToSign, $signingKey)
+            $signature
         );
 
         $request->setHeader('authorization', $authorizationHeader);
     }
 
-    protected function getCanonicalizedBody(Request $request): string
+    private function getContentLength(Request $request): ?int
     {
-        return hash('sha256', $request->getBody());
+        if ($request->hasHeader('content-length')) {
+            return (int) ((array) $request->getHeader('content-length'))[0];
+        }
+
+        $body = $request->getBody();
+        if (\is_string($body)) {
+            return \strlen($body);
+        }
+
+        if (\is_resource($body)) {
+            return fstat($body)['size'] ?? null;
+        }
+
+        return null;
     }
 
-    protected function getMd5Body(Request $request): string
+    private function prepareBody(Request $request, string $amzDate, string $credentialScope, string &$signature, string $signingKey): void
     {
-        return base64_encode(hash('md5', $request->getBody(), true));
+        $body = $request->getBody();
+
+        // we can't manage signature of undefined length closure
+        $contentLength = $this->getContentLength($request);
+        if (null === $contentLength && \is_callable($body)) {
+            $buffer = '';
+            while (true) {
+                if (!\is_string($data = $body(self::CHUNK_SIZE))) {
+                    throw new TransportException(sprintf('The return value of the "body" option callback must be a string, %s returned.', \gettype($data)));
+                }
+
+                if ('' == $data) {
+                    break;
+                }
+
+                $buffer .= $data;
+            }
+
+            $request->setBody($body = $buffer);
+        }
+
+        if (\is_string($body)) {
+            $request->setHeader('x-amz-content-sha256', hash('sha256', $body));
+            $request->setHeader('content-length', $contentLength);
+
+            return;
+        }
+
+        $streamReader = null;
+        if (\is_callable($body)) {
```

Comment:
This will only happen if body is a callable and the content-length header was defined, right?

Reply:
yes, if the size is unknown (not sent via header, and is not a stream), then the method tries first to convert the stream/callable into a string. And the follow the old way